### PR TITLE
Address State field drop down is ordered by ID, instead of State Name

### LINKED ISSUE #18731: Address State field drop down is ordered by ID, instead of State Name
#### Describe the bug

Anywhere where Address field is displayed (in client are or during checkout), the drop down with States orders values by database ID and not name field.
This makes the newly added States appear last in the list, thus confusing customer who will not see his State unless he will scroll to the bottom of the list.

#### Expected behavior

State field in address forms should be ordered by Name field in ps_state table, or whatever prefix is used.

#### Steps to Reproduce

1. From Client Area, go to Addresses
2. Edit or create a new address
3. State field (not visible by default but it can be added) will have values listed by ID.

**Screenshots**

Untitled4
Untitled5

#### Additional information

* PrestaShop version: Tested with 1.7.6.4 and 1.7.6.5

### PR DESCRIPTION
**Pull request by PululuK**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop 
| Description?  | Anywhere where Address field is displayed (in client are or during checkout), the drop down with States orders values by database ID and not name field.This makes the newly added States appear last in the list, thus confusing customer who will not see his State unless he will scroll to the bottom of the list
| Type?         | refacto 
| Category?     | CO 
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #18731.
| How to test?  | See #18731 by @vitciobanu.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/18736)

### CODE REVIEW

**Comment by Progi1984** [comment]
@PululuK Could you fix Coding Standards on your PR : `./vendor/bin/php-cs-fixer fix --dry-run --diff` ? Thanks :)

**Review by PierreRambaud — DISMISSED** [review]
You duplicate the getStatesByIdCountry logic. Feel free to improve the original method to allow the  sort by specific key instead.

**Comment by PululuK** [comment]
> You duplicate the getStatesByIdCountry logic. Feel free to improve the original method to allow the sort by specific key instead.

@PierreRambaud I thought about it, but I was afraid of the impact in other places where it is used ... I will do that then!

I like the idea thanks man !

**Comment by vitciobanu** [comment]
Guys, I'm impressed by your quick work on this _minor_ issue! Thank you!

I'm new to PrestaShop and issue resolutions here, so please let me know if you need me to download a new build to retest or something else.

**Comment by PierreRambaud** [comment]
> > You duplicate the getStatesByIdCountry logic. Feel free to improve the original method to allow the sort by specific key instead.
> 
> @PierreRambaud I thought about it, but I was afraid of the impact in other places where it is used ... I will do that then!
> 
> I like the idea thanks man !

Something like this, but I did not test it:
```php
    /**
     * Get states by Country ID.
     *
     * @param int $idCountry Country ID
     * @param bool $active true if the state must be active
     * @param string $orderBy Field to order result
     *
     * @return array|false|mysqli_result|PDOStatement|resource|null
     */
    public static function getStatesByIdCountry($idCountry, $active = false, $orderBy = null)
    {
        if (empty($idCountry)) {
            die(Tools::displayError());
        }

        $sql = 'SELECT * ' .
			'FROM `' . _DB_PREFIX_ . 'state` s ' .
			'WHERE s.`id_country` = ' . (int) $idCountry .
             (
                 $active ? ' AND s.active = 1' : ''
             );

        if ($orderBy !== null) {
            $sql .= 'ORDER BY s.`' . bqSQL($orderBy) . '`';
        }

        return Db::getInstance()->executeS(
            $sql
        );
    }

```

**Comment by Progi1984** [comment]
> > You duplicate the getStatesByIdCountry logic. Feel free to improve the original method to allow the sort by specific key instead.
> 
> @PierreRambaud I thought about it, but I was afraid of the impact in other places where it is used ... I will do that then!
> 
> I like the idea thanks man !

Use parameters for `getStatesByIdCountry` and let the initial behavior as default behavior

File: classes/State.php [hunk 1]
```diff
@@ -206,18 +206,22 @@ public function countUsed()
...
-			SELECT *
-			FROM `' . _DB_PREFIX_ . 'state` s
-			WHERE s.`id_country` = ' . (int) $idCountry . ($active ? ' AND s.active = 1' : '')
-        );
+        $sql = new DbQuery();
+        $sql->select('*');
+        $sql->from('state', 's');
+        $sql->where('s.id_country = ' . (int) $idCountry . ($active ? ' AND s.active = 1' : ''));
+
+        if (array_key_exists($orderBy, self::$definition['fields'])) {
```
Comment (PierreRambaud) [inline classes/State.php]:
```suggestion
        if (array_key_exists($orderBy, static::$definition['fields'])) {
```
If someone already override this class :sweat_smile:

File: classes/State.php [hunk 2]
```diff
@@ -206,18 +206,22 @@ public function countUsed()
      *
```
Comment (Progi1984) [inline classes/State.php:208]: Miss only $orderBy and it's :ok_hand: for me.

File: classes/State.php [hunk 3]
```diff
@@ -203,21 +203,26 @@ public function countUsed()
...
      * @param int $idCountry Country ID
      * @param bool $active true if the state must be active
+     * @param string $orderBy order by field
      *
      * @return array|false|mysqli_result|PDOStatement|resource|null
      */
-    public static function getStatesByIdCountry($idCountry, $active = false)
+    public static function getStatesByIdCountry($idCountry, $active = false, $orderBy = null)
```
Comment (jolelievre) [inline classes/State.php]: The only issue is you can't define the order way ASC or DESC?
Should we add another parameter, or use a convention like in the webservice `name_ASC`?
Or do you think it's not useful at all?

Comment (PululuK) [reply classes/State.php]: Very good question. I can add it in my PR if you allow me !
I would like to have the opinion of another developer
Thank you so much @jolelievre 

ping @PierreRambaud  @Progi1984

Comment (Progi1984) [reply classes/State.php]: Good idea @jolelievre

**Review by Progi1984 — DISMISSED** [review]

**Review by matks — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @PululuK,

In the FO, it is OK
image

But in the BO, it is NOK
image

Thanks to check and feedback.

**Comment by PululuK** [comment]
> Hi @PululuK,
> 
> In the FO, it is OK
> But in the BO, it is NOK

Thank you very much @khouloudbelguith ! but ... where exactly in BO ?

**Comment by khouloudbelguith** [comment]
@PululuK, in the BO => Customers => Addresses => try to add a new address and select a country which contains states for example Spain
image

Thanks!

**Comment by PululuK** [comment]
> @PululuK, in the BO => Customers => Addresses => try to add a new address and select a country which contains states for example Spain
> Thanks!

Done ! Thank you @khouloudbelguith

**Review by Progi1984 — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @PululuK,

It is OK :heavy_check_mark: 
**BO**
image
**FO**
image

Thanks!

**Comment by Progi1984** [comment]
Thanks @PululuK

**Pull request merged**